### DISCUSSION TITLE
Labs Feedback: @lit-labs/testing

### DISCUSSION THREAD

**Discussion by augustjk** [lit/lit#3359] [Labs Feedback]
The `@lit-labs/testing` package contains testing utilities for Lit including generating server-side rendered fixtures.

- Status: **Under development**

- Issues: [[labs/testing]](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Ftesting%5D+in%3Atitle)
- Docs: [README](https://github.com/lit/lit/tree/main/packages/labs/testing#readme)
- npm: [@lit-labs/testing](https://www.npmjs.com/package/@lit-labs/testing)

Please try `@lit-labs/testing` and tell us about your experience. Leave comments with your thoughts; highlight what issues or missing features you're blocked on; give a link to a project that uses it; etc.

And so we can have some numbers to look at, you can add reactions to this comment to indicate your experience with this package:

🚀 I have tried this package and it works great. I need it. Ship it!
👍 I have tried this package and it seems like a good idea.
😕 I have tried this package and would use it, but it has serious bugs or missing features.
👀 I am interested, but haven't tried the package yet.
👎 I'm not interested in this package.

Note: other reactions like 😄, 🎉, and ❤️ don't mean anything as part of this scheme. You probably mean, and please use, 🚀, 👍, or 👀.

**Comment by kyubisation** [comment]
Thank you very much for this library. It is a very nice convenience. 👍 

We are currently using this library together with Web Test Runner.
From my understanding, the `ssrNonHydratedFixture` case cannot work properly as it is, as importing the fixtures will always load `@lit-labs/ssr-client/lit-element-hydrate-support.js`, which will mutate the `LitElement` (and `ReactiveElement`) prototype, which in turn would prevent `connectedCallback` (including added `ReactiveController` instances) from being executed.
As I understand it, `ssrNonHydratedFixture` should not have `@lit-labs/ssr-client/lit-element-hydrate-support.js` imported.

Or am I misunderstanding the usage?

**Reply by augustjk** [reply]
Thanks for trying it out!

That is the intended behavior of `ssrNonHydratedFixture`. It simulates aving a server rendered component loaded to the page with declarative shadow DOM with no component definition JavaScript having been loaded.

It'd imagine it's possible to have components that are purely for presentation purposes that won't need hydration to enable interactivity, but nice to still make it a custom element for scoped styling and such. It would mainly be for those.

It's also possible users would intentionally not hydrate the component with the `defer-hydration` attribute (which is how this fixture util does it), for example, in order to add some data to it before hydration and first update.

Or if there's some kind of progressive enhancement that the component should have some behavior before hydration, that's another reason to use this fixture to test that.

In any case, it won't be possible to selectively import the hydrate support in one fixture and not the other since if you have a mix of fixtures you can't un-import it just for the execution of the non-hydrated ones. The hydrate support is transparent for any non server rendered Lit components anyway and only stops the "upgrade" from happening for those with `defer-hydration`.

**Reply by kyubisation** [reply]
Thank you for the in-depth explanation.
This purpose was not clear to me from reading https://github.com/lit/lit/blob/main/packages/labs/testing/README.md.
With this understanding, the package works as intended and we need to adapt our usage.
